### PR TITLE
HIA-657: 🎨 Convert filter to kotlin

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 1]
```diff
@@ -611,15 +611,15 @@ public Page<CaseNote> getOffenderCaseNotes(@PathVariable("offenderNo") @Paramete
...
-        final var caseNoteFilter = CaseNoteFilter.builder()
-            .type(type)
-            .subType(subType)
-            .typesSubTypes(typeSubTypes)
-            .prisonId(prisonId)
-            .startDate(from)
-            .endDate(to)
-            .bookingId(latestBookingByOffenderNo.getBookingId())
-            .build();
+        final var caseNoteFilter = new CaseNoteFilter(
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:614]: will look nicer when the resource is converted to kotlin too

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import jakarta.persistence.criteria.Predicate
+import jakarta.persistence.criteria.Root
+import org.springframework.data.jpa.domain.Specification
+import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderCaseNote
+import uk.gov.justice.hmpps.prison.service.BadRequestException
+import uk.gov.justice.hmpps.prison.service.QueryParamHelper
+import java.time.LocalDate
+
+class CaseNoteFilter(
+  val bookingId: Long,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt:15]: I've made the first field here not nullable as `bookingService.getLatestBookingByOffenderNo` throws an exception if the offender isn't found and hence the booking

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package uk.gov.justice.hmpps.prison.repository.jpa.repository
+
+import com.google.common.collect.ImmutableList
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt:3]: only just spotted this, will remove in next PR

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  type: String? = null,
+  subType: String? = null,
+  typesSubTypes: List<String>? = null,
+) : Specification<OffenderCaseNote> {
+
+  private val typesAndSubTypes = if (!type.isNullOrEmpty()) {
+    if (!typesSubTypes.isNullOrEmpty()) {
+      throw BadRequestException("Both type and typesSubTypes are set, please only use one to filter.")
+    }
+    if (subType.isNullOrEmpty()) listOf(type) else listOf("$type+$subType")
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt:28]: simplify original java by adding the `type` and `subType` parameters to the list

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceIntTest.kt [hunk 5]
```diff
@@ -209,7 +209,7 @@ class OffenderResourceIntTest : ResourceTest() {
       )
       val error = response.body!!
       assertThat(response.statusCode.value()).isEqualTo(400)
-      assertThat(error.userMessage).contains("Both type and typesAndSubTypes are set, please only use one to filter.")
+      assertThat(error.userMessage).contains("Both type and typesSubTypes are set, please only use one to filter.")
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderResourceIntTest.kt:212]: error message changed to match input parameter

**Review by CslBing — APPROVED** [review]
Looks good

**Pull request merged**